src/python/pants/backend/docker/rules_context.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from dataclasses import dataclass
+from itertools import chain
+from typing import Tuple
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.core.target_types import FilesSources, ResourcesSources
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import AddPrefix, Digest, MergeDigests
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    Dependencies,
+    DependenciesRequest,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    Sources,
+    Target,
+    Targets,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DockerBuildContext:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class DockerBuildContextRequest:
+    address: Address
+    context_root: str
+    targets: Tuple[Target, ...]
+
+
+@rule
+async def create_docker_build_context(request: DockerBuildContextRequest) -> DockerBuildContext:
+    # Get all sources, going into the build context.
+    sources_request = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[t.get(Sources) for t in request.targets],
+            for_sources_types=(DockerImageSources, FilesSources, ResourcesSources),
```

Comment:
The source for `DockerImageSources` is the Dockerfile itself! I don't think that counts as part of the context, because it isn't going to be copied into the image? It should be handled out of band.

Reply:
No, not into the image, but it needs to be part of the build context, otherwise it won't end up with the docker daemon building the image (so, wouldn't know what to build)